bindgen/features.rs

```diff
@@ -4,270 +4,227 @@
-            => Stable_1_71 => 1.71;
-            /// Nightly rust
-            ///  * `thiscall` calling convention ([Tracking issue](https://github.com/rust-lang/rust/issues/42202))
-            ///  * `vectorcall` calling convention (no tracking issue)
-            => Nightly => nightly;
-        );
+impl PartialOrd for RustTarget {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
 }
 
-rust_target_base!(rust_target_def);
-rust_target_base!(rust_target_values_def);
+impl Ord for RustTarget {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self.minor(), other.minor()) {
+            (Some(a), Some(b)) => a.cmp(&b),
+            (Some(_), None) => Ordering::Less,
+            (None, Some(_)) => Ordering::Greater,
+            (None, None) => Ordering::Equal,
+        }
+    }
+}
 
-/// Latest stable release of Rust
-pub const LATEST_STABLE_RUST: RustTarget = RustTarget::Stable_1_71;
+impl FromStr for RustTarget {
+    type Err = io::Error;
 
-/// Create RustFeatures struct definition, new(), and a getter for each field
-macro_rules! rust_feature_def {
-    (
-        $( $rust_target:ident {
-            $( $( #[$attr:meta] )* => $feature:ident; )*
-        } )*
-    ) => {
-        /// Features supported by a rust target
-        #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
-        #[allow(missing_docs)] // Documentation should go into the relevant variants.
-        pub(crate) struct RustFeatures {
-            $( $(
-                $(
-                    #[$attr]
-                )*
-                pub $feature: bool,
-            )* )*
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "nightly" {
+            return Ok(Self::Nightly);
         }
 
-        impl RustFeatures {
-            /// Gives a RustFeatures struct with all features disabled
-            fn new() -> Self {
-                RustFeatures {
-                    $( $(
-                        $feature: false,
-                    )* )*
+        if let Some(("1", str_minor)) = s.split_once('.') {
+            if let Ok(minor) = str_minor.parse::<u64>() {
+                for (target, target_minor) in Self::stable_releases() {
+                    if minor == target_minor {
+                        return Ok(target);
+                    }
                 }
             }
         }
 
-        impl From<RustTarget> for RustFeatures {
-            fn from(rust_target: RustTarget) -> Self {
-                let mut features = RustFeatures::new();
-
-                $(
-                if rust_target >= RustTarget::$rust_target {
-                    $(
-                    features.$feature = true;
-                    )*
-                }
-                )*
-
-                features
-            }
-        }
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Got an invalid rust target. Accepted values are of the form \"1.0\" or \"nightly\"."
```

Comment:
```suggestion
            "Got an invalid Rust target. Accepted values are of the form \"1.71\" or \"nightly\"."
```
- consistent to upper-case `rust`
- choose a more realistic example for version number